lib/tracing-limit/src/lib.rs

```diff
@@ -10,199 +9,204 @@ use std::{
-
-                        let meta = event.metadata();
-
-                        let mut visitor = FmtVisitor::default();
-                        event.record(&mut visitor);
-
-                        let message = if let Some(message) = &visitor.message {
-                            &message
+    fn enabled(&self, metadata: &Metadata, ctx: Context<S>) -> bool {
+        if is_limited(metadata) {
+            let id = metadata.callsite();
+            let events = self.events.read().expect("lock poisoned!");
+
+            // check if the event exists within the map, if it does
+            // that means we are currently rate limiting it.
+            if let Some(state) = events.get(&id) {
+                let ts = SystemTime::now()
+                    .duration_since(UNIX_EPOCH)
+                    .unwrap()
+                    .as_secs() as usize;
+
+                if ts - state.start < state.limit {
+                    if state.count.load(Ordering::SeqCst) == 1 {
+                        let message = if let Some(event_message) = &state.message {
+                            format!("{:?} is being rate limited.", event_message)
                         } else {
-                            "unknown event"
+                            format!("unknown message is being rate limited.")
                         };
 
-                        // We need to replicate the way that fmt logs events
-                        // because we currently can not create new fresh events.
-                        println!(
-                            "{} {} {:?} {:?} logs were rate limited.",
-                            FmtLevel(meta.level()),
-                            meta.target(),
-                            local_count,
-                            message
-                        );
-
-                        return;
+                        self.create_event(&id, &ctx, message);
+                    }
+
+                    let prev = state.count.fetch_add(1, Ordering::SeqCst);
+
+                    if prev == 0 {
+                        return true;
                     }
                 } else {
                     drop(events);
-                    let count = AtomicUsize::new(1);
-                    let ts = AtomicUsize::new(ts as usize);
-                    let mut map = self.events.write().unwrap();
-                    map.insert(id, (count, ts));
+
+                    let mut events = self.events.write().expect("lock poisoned!");
+
+                    if let Some(state) = events.remove(&id) {
+                        let message = if let Some(event_message) = &state.message {
+                            format!(
+                                "{:?} {:?} events were rate limited.",
+                                state.count, event_message
+                            )
+                        } else {
+                            format!("{:?} unknown messages were rate limited.", state.count)
```

Comment:
TIOLI: you could use the event's name (`event.metadata().name()`) in the case when it doesn't have a message.